File: libpandadna/DNAStreet.py
import DNANode

from common import *

class DNAStreet(DNANode.DNANode):
    COMPONENT_CODE = 19

    def __init__(self, name):
        DNANode.DNANode.__init__(self, name)

        self.code = ''
        self.streetTexture = ''
        self.sideWalkTexture = ''
        self.curbTexture = ''
        self.streetColor = (1, 1, 1, 1)
        self.sideWalkColor = (1, 1, 1, 1)
        self.curbColor = (1, 1, 1, 1)

    def makeFromDGI(self, dgi):
        DNANode.DNANode.makeFromDGI(self, dgi)
        
        self.code = dgi_extract_string8(dgi)
        self.streetTexture = dgi_extract_string8(dgi)
        self.sidewalkTexture = dgi_extract_string8(dgi)
        self.curbTexture = dgi_extract_string8(dgi)

        self.streetColor = dgi_extract_color(dgi)
        self.sideWalkColor = dgi_extract_color(dgi)
        self.curbColor = dgi_extract_color(dgi)
        
    def getTexture(self, texture, store):
        if not texture:
            return
            
        tex = store.findTexture(texture)
        if tex is None:
            self.raiseCodeNotFound('texture: ' + texture)
            
        return tex
        
    def traverse(self, np, store):
        node = store.findNode(self.code)
        if node is None:
            self.raiseCodeNotFound()
            
        _np = node.copyTo(np)
        
        node.setName(self.name)
        
        streetTexture = self.getTexture(self.streetTexture, store)
        sidewalkTexture = self.getTexture(self.sidewalkTexture, store)
        curbTexture = self.getTexture(self.curbTexture, store)

        streetNode = _np.find('**/*_street')
        sidewalkNode = _np.find('**/*_sidewalk')
        curbNode = _np.find('**/*_curb')

        if not streetNode.isEmpty() and streetTexture:
            streetNode.setTexture(streetTexture, 1)
            streetNode.setColorScale(self.streetColor)
            
        if not sidewalkNode.isEmpty() and sidewalkTexture:
            sidewalkNode.setTexture(sidewalkTexture, 1)
            sidewalkNode.setColorScale(self.sideWalkColor)
            
        if not curbNode.isEmpty() and curbTexture:
            curbNode.setTexture(curbTexture, 1)
            curbNode.setColorScale(self.curbColor)

        _np.setPosHprScale(self.pos, self.hpr, self.scale)

File: src/components/DNAStreet.cxx
#include "DNAStreet.h"

TypeHandle DNAStreet::_type_handle;

DNAStreet::DNAStreet(const std::string& name): DNANode(name), m_code(""),
                                               m_street_texture(""),
                                               m_sidewalk_texture(""),
                                               m_curb_texture(""),
                                               m_street_color(LVecBase4f(1)),
                                               m_sidewalk_color(LVecBase4f(1)),
                                               m_curb_color(LVecBase4f(1)),
                                               m_set_texture_count(0),
                                               m_set_color_count(0)
{
}

DNAStreet::~DNAStreet()
{
}

void DNAStreet::make_from_dgi(DatagramIterator& dgi, DNAStorage* store)
{
    DNANode::make_from_dgi(dgi, store);
    m_code = dgi.get_string();
    m_street_texture = dgi.get_string();
    m_sidewalk_texture = dgi.get_string();
    m_curb_texture = dgi.get_string();
    dgi_extract_color(dgi, m_street_color);
    dgi_extract_color(dgi, m_sidewalk_color);
    dgi_extract_color(dgi, m_curb_color);
}

void DNAStreet::traverse(NodePath& np, DNAStorage* store)
{
    NodePath result = store->find_node(m_code);
    if (result.is_empty())
    {
        raise_code_not_found();
        return;
    }

    NodePath _np = result.copy_to(np);
    _np.set_name(m_name);

    PT(Texture) street_texture = get_texture(m_street_texture, store);
    PT(Texture) sidewalk_texture = get_texture(m_sidewalk_texture, store);
    PT(Texture) curb_texture = get_texture(m_curb_texture, store);

    NodePath street_node = _np.find("**/*_street");
    NodePath sidewalk_node = _np.find("**/*_sidewalk");
    NodePath curb_node = _np.find("**/*_curb");

    if (!street_node.is_empty() && street_texture)
    {
        street_node.set_texture(street_texture, 1);
        street_node.set_color_scale(m_street_color);
    }

    if (!sidewalk_node.is_empty() && sidewalk_texture)
    {
        sidewalk_node.set_texture(sidewalk_texture, 1);
        sidewalk_node.set_color_scale(m_sidewalk_color);
    }

    if (!curb_node.is_empty() && curb_node)
    {
        curb_node.set_texture(curb_texture, 1);
        curb_node.set_color_scale(m_curb_color);
    }

    _np.set_pos_hpr_scale(m_pos, m_hpr, m_scale);
}

void DNAStreet::set_texture(const std::string& texture)
{
    if (m_set_texture_count == 0)
        set_street_texture(texture);

    else if (m_set_texture_count == 1)
        set_sidewalk_texture(texture);

    else if (m_set_texture_count == 2)
        set_curb_texture(texture);

    else
    {
        dna_cat.warning() << "tried to add a 4th texture to DNAStreet!" << std::endl;
        return;
    }

    m_set_texture_count++;
}

void DNAStreet::set_color(LVecBase4f color)
{
    if (m_set_color_count == 0)
        set_street_color(color);

    else if (m_set_color_count == 1)
        set_sidewalk_color(color);

    else if (m_set_color_count == 2)
        set_curb_color(color);

    else
    {
        dna_cat.warning() << "tried to add a 4th color to DNAStreet!" << std::endl;
        return;
    }

    m_set_color_count++;
}

PT(Texture) DNAStreet::get_texture(const std::string& texture, DNAStorage* store)
{
    if (!texture.size())
        return nullptr;

    PT(Texture) tex = store->find_texture(texture);
    if (!tex)
    {
        raise_code_not_found(texture.c_str());
        return nullptr;
    }

    return tex;
}
